client/verta/tests/test_endpoint/test_endpoint.py

```diff
@@ -600,3 +600,23 @@ def test_update_twice(self, client, registered_model, created_endpoints):
         endpoint.update(new_model_version, DirectUpdateStrategy(), wait=True)
         test_data = np.random.random((4, 12))
         assert np.array_equal(endpoint.get_deployed_model().predict(test_data), new_classifier.predict(test_data))
+
+    def test_update_build_error(self, client, created_endpoints, experiment_run, model_for_deployment, model_version):
+        model_version.log_model(model_for_deployment['model'], custom_modules=[])
+        model_version.log_environment(Python(requirements=["scikit-learn"]))
+
+        experiment_run.log_model(model_for_deployment['model'], custom_modules=[])
+        experiment_run.log_requirements(['blahblahblah==3.6.0'])
+
+        path = verta._internal_utils._utils.generate_default_name()
+        endpoint = client.set_endpoint(path)
+        created_endpoints.append(endpoint)
+
+        update_status = endpoint.update(model_version, DirectUpdateStrategy(), True) # this should work
+        assert update_status["status"] == "active"
+
+        with pytest.raises(RuntimeError) as excinfo:
+            endpoint.update(experiment_run, DirectUpdateStrategy(), True) # this should fail, and not take forever!
+
+        excinfo_value = str(excinfo.value).strip()
+        assert "Could not find a version that satisfies the requirement blahblahblah==3.6.0" in excinfo_value
```

Comment:
- No need to test the successful case; other tests already have that covered (and the Client test suite is getting strapped for runtime 😆)
- As you noted, a bug in the Client would result in this test taking forever, rather than failing. We should use a loop that checks the endpoint status every ~5 seconds instead of `wait=True`, and fail the test if it exceeds a time limit.

Reply:
* I can remove the assertion for successful case, but I thought the "taking forever" bug only happen if we first updated successfully, then update again with a fail build?

Reply:
@nhatsmrt I had forgotten about "taking forever" only occurring after a successful initial update.

Would you know if that's still the case? I feel like I've encountered "taking forever" recently even without a successful initial update, but I could be mistaken

Reply:
Hmm, for me it only fails when there is a successful initial update.

Reply:
@convoliution I rmb replying to this hmmm...
Anyway for me the "taking forever" bug only happens with a successful initial update.

Reply:
@nhatsmrt Alright. In that case, do you think we could graft this "taking forever" case to the end of an existing endpoint update test?

Reply:
@convoliution and remove this test?

Reply:
Yes.

Reply:
@convoliution Hmm about your other suggestion:

"We should use a loop that checks the endpoint status every ~5 seconds instead of wait=True, and fail the test if it exceeds a time limit."

The point of this test is to check if endpoint.update(..., wait=True) does not exceed the time limit right? If we check the endpoint's status (which gets the stage's status), it will never be ready (hence the need for this PR). If we check the actual build status, then it just defeats the point of the test, because we basically just repeat the implementation. Also if the client is buggy and it does take forever in the case of "wait=True", we won't detect it anyway by checking via build anyway (since that works properly all the time).

Reply:
I wonder if pytest has a mechanism to just time out a block of code/a test if it takes too long 🤔

Reply:
@nhatsmrt
>it just defeats the point of the test

Hah! You're right. My suggestions so far have all been misses. But ultimately we have these key constraints:

- this test *must* not literally take forever to execute, because it'll block the rest of the suite
- this test *should* not require additional pytest plugins (e.g. time limit on a test), because plugins caused various issues in the past, but if the suite still works then sure
- this test *should* be attached to the end of an existing one, because the suite is approaching the limits of its runtime

I'm not sure what the solution will be (I'm tied up at the moment with a major bug) so I'll leave it to you to try to find a solution! 🙇